From 39612058971cff01e1f65fa475685dfc19a47234 Mon Sep 17 00:00:00 2001
From: Roger Leigh <rleigh@debian.org>
Date: Mon, 25 Jun 2012 20:30:46 +0100
Subject: [PATCH 1/2] debian: Correct library versioning and linking

- Explictly cater for direct linkage of all undefined symbols.
  Indirect linking does not work with current toolchains, which
  require first-order linkage of all undefined symbols.
- Disable use of INTERFACE_AGE in library versions.
- Correct unintended ABI version bump.
  This sets the soname of libgutenprint to 2.1.0 instead of 3.0.0.

This has been committed upstream for 5.2.9, but this corrects
the ABI versioning error in libgutenprint2 for 5.2.8 in Debian.
---
 configure.ac                |   22 +++++++++++-----------
 src/gimp2/Makefile.am       |    4 ++--
 src/testpattern/Makefile.am |    2 +-
 3 files changed, 14 insertions(+), 14 deletions(-)

diff --git a/configure.ac b/configure.ac
index 0dc3241..f268689 100644
--- a/configure.ac
+++ b/configure.ac
@@ -23,13 +23,12 @@ dnl
 dnl Since the last release:
 dnl 1. if only source code (not the interface) has changed, set
 dnl      GUTENPRINT_MICRO_VERSION += 1;
-dnl      GUTENPRINT_INTERFACE_AGE += 1;
 dnl 2. if any functions have been added, removed, or changed, set
-dnl      GUTENPRINT_INTERFACE_AGE = 0;
 dnl      GUTENPRINT_CURRENT_INTERFACE += 1;
 dnl 3. if interfaces have been added, set
 dnl      GUTENPRINT_BINARY_AGE += 1;
-dnl 4. if interfaces have been removed, set
+dnl    or, if interfaces have been removed, set
+dnl      GUTENPRINT_CURRENT_INTERFACE -= GUTENPRINT_BINARY_AGE
 dnl      GUTENPRINT_BINARY_AGE = 0;
 dnl
 dnl For more detailed information, see the libtool info documentation.
@@ -40,10 +39,8 @@ pushdef([GUTENPRINT_MINOR_VERSION],     [2])
 pushdef([GUTENPRINT_MICRO_VERSION],     [8])
 pushdef([GUTENPRINT_EXTRA_VERSION],     [])
 pushdef([GUTENPRINT_CURRENT_INTERFACE], [3])
-pushdef([GUTENPRINT_INTERFACE_AGE],     [0])
-pushdef([GUTENPRINT_BINARY_AGE],        [0])
+pushdef([GUTENPRINT_BINARY_AGE],        [1])
 pushdef([GUTENPRINTUI2_CURRENT_INTERFACE], [1])
-pushdef([GUTENPRINTUI2_INTERFACE_AGE],     [0])
 pushdef([GUTENPRINTUI2_BINARY_AGE],        [0])
 pushdef([GUTENPRINT_VERSION], GUTENPRINT_MAJOR_VERSION.GUTENPRINT_MINOR_VERSION.GUTENPRINT_MICRO_VERSION[]GUTENPRINT_EXTRA_VERSION)
 
@@ -76,21 +73,19 @@ Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
 [GUTENPRINT_MICRO_VERSION]=GUTENPRINT_MICRO_VERSION
 [GUTENPRINT_EXTRA_VERSION]=GUTENPRINT_EXTRA_VERSION
 [GUTENPRINT_CURRENT_INTERFACE]=GUTENPRINT_CURRENT_INTERFACE
-[GUTENPRINT_INTERFACE_AGE]=GUTENPRINT_INTERFACE_AGE
+[GUTENPRINT_INTERFACE_AGE]=0
 [GUTENPRINT_BINARY_AGE]=GUTENPRINT_BINARY_AGE
 [GUTENPRINT_VERSION]=GUTENPRINT_VERSION
 [GUTENPRINTUI2_CURRENT_INTERFACE]=GUTENPRINTUI2_CURRENT_INTERFACE
-[GUTENPRINTUI2_INTERFACE_AGE]=GUTENPRINTUI2_INTERFACE_AGE
+[GUTENPRINTUI2_INTERFACE_AGE]=0
 [GUTENPRINTUI2_BINARY_AGE]=GUTENPRINTUI2_BINARY_AGE
 popdef([GUTENPRINT_MAJOR_VERSION])
 popdef([GUTENPRINT_MINOR_VERSION])
 popdef([GUTENPRINT_MICRO_VERSION])
 popdef([GUTENPRINT_EXTRA_VERSION])
 popdef([GUTENPRINT_CURRENT_INTERFACE])
-popdef([GUTENPRINT_INTERFACE_AGE])
 popdef([GUTENPRINT_BINARY_AGE])
 popdef([GUTENPRINTUI2_CURRENT_INTERFACE])
-popdef([GUTENPRINTUI2_INTERFACE_AGE])
 popdef([GUTENPRINTUI2_BINARY_AGE])
 
 AC_SUBST([GUTENPRINT_MAJOR_VERSION])
@@ -738,7 +733,11 @@ AC_CHECK_LIB(dl, dlopen, [DLOPEN_LIBS="-ldl"])
 
 AC_CHECK_LIB(m,pow,
              GUTENPRINT_LIBDEPS="${GUTENPRINT_LIBDEPS} -lm"
-             gutenprint_libdeps="${gutenprint_libdeps} -lm")
+             gutenprint_libdeps="${gutenprint_libdeps} -lm"
+             GUTENPRINTUI2_LIBDEPS="${GUTENPRINT_LIBDEPS} -lm"
+             gutenprintui2_libdeps="${gutenprint_libdeps} -lm"
+	     LIBM=-lm
+)
 
 STP_CUPS_LIBS
 
@@ -950,6 +949,7 @@ AC_SUBST(gutenprintui2_cflags)
 gutenprintui2_libs="${GUTENPRINTUI2_LIBS} ${gutenprintui2_libdeps}"
 AC_SUBST(gutenprintui2_libs)
 AC_SUBST(gutenprintui2_libdeps)
+AC_SUBST(LIBM)
 AC_SUBST(LIBREADLINE_DEPS)
 AC_SUBST(MAINTAINER_CFLAGS)
 AC_SUBST(PLUG_IN_PATH)
diff --git a/src/gimp2/Makefile.am b/src/gimp2/Makefile.am
index 489e6aa..a23cd2c 100644
--- a/src/gimp2/Makefile.am
+++ b/src/gimp2/Makefile.am
@@ -46,8 +46,8 @@ COMMON_PRINT_SOURCES = \
 
 
 print_SOURCES = $(COMMON_PRINT_SOURCES) print-print.c
-print_LDADD = $(GIMP2_LIBS) $(GUTENPRINTUI2_LIBS)
-print_DEPENDENCIES = $(GUTENPRINTUI2_LIBS) $(GUTENPRINT_LIBS)
+print_LDADD = $(GIMP2_LIBS) $(GUTENPRINT_LIBS) $(GUTENPRINTUI2_LIBS)
+print_DEPENDENCIES = $(GUTENPRINT_LIBS) $(GUTENPRINTUI2_LIBS)
 
 gutenprint_SOURCES = $(COMMON_PRINT_SOURCES) print-gutenprint.c
 gutenprint_LDADD = $(print_LDADD)
diff --git a/src/testpattern/Makefile.am b/src/testpattern/Makefile.am
index 6d8434a..2c5bb68 100644
--- a/src/testpattern/Makefile.am
+++ b/src/testpattern/Makefile.am
@@ -41,7 +41,7 @@ AM_LFLAGS = -i
 AM_YFLAGS = -d
 
 testpattern_SOURCES = testpattern.c testpatterny.y testpatternl.l testpattern.h
-testpattern_LDADD = $(GUTENPRINT_LIBS)
+testpattern_LDADD = $(GUTENPRINT_LIBS) $(LIBM)
 
 testpatternl.o: testpatterny.o
 testpattern.o: testpatterny.o
